rustls-libssl/src/lib.rs

```diff
@@ -214,6 +215,167 @@ static TLS13_CHACHA20_POLY1305_SHA256: SslCipher = SslCipher {
+                + self.value.len()
+                + context_len.len()
+                + self.context.len(),
+        );
+        ret.extend_from_slice(SslSession::MAGIC);
+        ret.extend_from_slice(&expiry);
+        ret.extend_from_slice(&id_len);
+        ret.extend_from_slice(&self.id.0);
+        ret.extend_from_slice(&value_len);
+        ret.extend_from_slice(&self.value);
+        ret.extend_from_slice(&context_len);
+        ret.extend_from_slice(&self.context);
+        ret
+    }
+
+    /// Decodes from the front of `slice`.  Returns the remainder.
+    pub fn decode(slice: &[u8]) -> Option<(Self, &[u8])> {
+        fn split_at(slice: &[u8], mid: usize) -> Option<(&[u8], &[u8])> {
+            if mid <= slice.len() {
+                Some(slice.split_at(mid))
+            } else {
+                None
+            }
+        }
+
+        fn slice_to_usize(slice: &[u8]) -> usize {
+            // unwrap: `slice` must be `usize_len` in length
+            usize::from_le_bytes(slice.try_into().unwrap())
+        }
+
+        fn slice_to_u64(slice: &[u8]) -> u64 {
+            // unwrap: `slice` must be `u64_len` in length
+            u64::from_le_bytes(slice.try_into().unwrap())
+        }
+
+        let usize_len = mem::size_of::<usize>();
+        let u64_len = mem::size_of::<u64>();
+
+        let (magic, slice) = split_at(slice, SslSession::MAGIC.len())?;
+        if magic != SslSession::MAGIC {
+            return None;
+        }
+        let (expiry, slice) = split_at(slice, u64_len)?;
+        let (id_len, slice) = split_at(slice, usize_len)?;
+        let (id, slice) = split_at(slice, slice_to_usize(id_len))?;
+        let (value_len, slice) = split_at(slice, usize_len)?;
+        let (value, slice) = split_at(slice, slice_to_usize(value_len))?;
+        let (context_len, slice) = split_at(slice, usize_len)?;
+        let (context, slice) = split_at(slice, slice_to_usize(context_len))?;
+        let ret = Self {
+            id: SslSessionLookup(id.to_vec()),
+            value: value.to_vec(),
+            context: context.to_vec(),
+            expiry_time: cache::ExpiryTime(slice_to_u64(expiry)),
+        };
+        Some((ret, slice))
+    }
+
+    pub fn get_id(&self) -> &[u8] {
+        &self.id.0
+    }
+
+    pub fn expired(&self, at_time: cache::TimeBase) -> bool {
+        self.expiry_time.in_past(at_time)
+    }
+
+    pub fn older_than(&self, other: &Self) -> bool {
+        self.expiry_time.0 < other.expiry_time.0
+    }
+}
+
+impl cmp::PartialOrd<SslSession> for SslSession {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.id.cmp(&other.id))
+    }
+}
+
+impl cmp::Ord for SslSession {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.id.cmp(&other.id)
+    }
+}
+
+impl cmp::PartialEq<SslSession> for SslSession {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl cmp::Eq for SslSession {}
```

Comment:
Perhaps I misunderstood, but I think that import just allows (eg) `cmp::Ord` rather than `core::cmp::Ord`, but wouldn't allow just `Ord` (the import for that would be, `use core::cmp::{Ord, Ordering, PartialOrd, Eq, PartialEq};`). Generally I like this style for std/core imports because they tend to have lots of short/generic/clashing names like `Error` in there (`io::Error`, `fmt::Error`, etc.)

Reply:
Huh. What you're saying w.r.t the `cmp` import makes sense to me, but in practice I seem to be able to drop the `cmp::` prefix without error.

<details>
<summary>This diff builds for me:</summary>

```diff
diff --git a/rustls-libssl/src/lib.rs b/rustls-libssl/src/lib.rs
index 132386a..b279a7b 100644
--- a/rustls-libssl/src/lib.rs
+++ b/rustls-libssl/src/lib.rs
@@ -333,25 +333,25 @@ impl SslSession {
     }
 }
 
-impl cmp::PartialOrd<SslSession> for SslSession {
+impl PartialOrd<SslSession> for SslSession {
     fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
         Some(self.id.cmp(&other.id))
     }
 }
 
-impl cmp::Ord for SslSession {
+impl Ord for SslSession {
     fn cmp(&self, other: &Self) -> cmp::Ordering {
         self.id.cmp(&other.id)
     }
 }
 
-impl cmp::PartialEq<SslSession> for SslSession {
+impl PartialEq<SslSession> for SslSession {
     fn eq(&self, other: &Self) -> bool {
         self.id == other.id
     }
 }
 
-impl cmp::Eq for SslSession {}
+impl Eq for SslSession {}
 
 impl borrow::Borrow<SslSessionLookup> for Arc<SslSession> {
     fn borrow(&self) -> &SslSessionLookup {
@@ -360,7 +360,7 @@ impl borrow::Borrow<SslSessionLookup> for Arc<SslSession> {
 }
 
 impl fmt::Debug for SslSession {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
         f.debug_struct("SslSession")
             .field("id", &self.id)
             .field("expiry", &self.expiry_time)

```
</details>

Perhaps it's prelude related?

> Generally I like this style for std/core imports because they tend to have lots of short/generic/clashing names 

Fair! I don't feel strongly in either case, just a small mystery. Feel free to resolve with the code as-is :+1:

Reply:
Oh! Yes, that makes sense -- they're in the prelude.